src/core/protocol.ts

```diff
@@ -145,3 +32,83 @@ export class Protocol {
     return request;
   }
 }
+
+export class RedisProtocolError extends Error {
+
+  public static fromMessage(message: string) {
+    const space = message.indexOf(" ");
+    return new RedisProtocolError(message.slice(0, space), message.slice(space + 1));
+  }
+  constructor(name: string, message: string) {
+    super();
+    this.name = name;
+    this.message = message;
+  }
+}
+
+class ProtocolParser {
+  public static parse(raw: string): any[] {
```

Comment:
I just pushed this simplification to the `masterRegex`.  All my previous explanation still stands, this just tries to simplify the reusable stuff.

Reply:
An example:

`+OK\r\n-Error Some Error Message\r\n:-100\r\n:100\r\n$3\r\nfoo\r\n*2\r\n+SOMETHING\r\n$-1\r\n:42\r\n$10\r\nincomple`

`matchAll` would result in 9 matches, creating an array like (simplified):

```javascript
[
    { ... groups: {simple: "OK", error: undefined, int: undefined, bulk_n: undefined, bulk: undefined, array_n: undefined, array: undefined, null_string: undefined} ...},  
    { ... groups: { ... error: "Error Some Error" ... } ...},  
    { ... groups: { ... int: -100 ... } ...}
    { ... groups: { ... int: 100 ... } ...}
    { ... groups: { ... bulk_n: 3, bulk: "foo" ... } ...}
    { ... groups: { ... array_n: 2, array: null ... } ...}
    { ... groups: {simple: "SOMETHING" ... } ...}
    { ... groups: { ... null_string: -1} ...}
    { ... groups: { ... int: 42 ... } ...}
]
```

Which would then get aggregated to something like:

```javascript
[
    "OK",
    RedisProtocolError: {name: "Error", message: "Some Error"},
    -100,
    100,
    "foo",
    [ 
        "SOMETHING",
        null,
    ],
    42,
]
```